sycl/include/CL/sycl/intel/pipes.hpp

```diff
@@ -84,6 +84,112 @@ template <class name, class dataT, int32_t min_capacity = 0> class pipe {
 #endif // __SYCL_DEVICE_ONLY__
 };
 
+// IO pipes that provide interface to connect with hardware peripheral.
+// Their name aliases are defined in vendor-provided header, below you can see
+// an example of this header. There are defined aliases to ethernet_read_pipe
+// and ethernet_write_pipe that users can use in their code to connect with
+// HW peripheral.
+/* namespace intelfpga {
+template <int32_t ID>
+struct ethernet_pipe_id {
+  static constexpr int32_t id = ID;
+};
+
+template <class dataT, size_t min_capacity>
+using ethernet_read_pipe =
+  kernel_readable_io_pipe<ethernet_pipe_id<0>, dataT, min_capacity>;
+
+template <class dataT, size_t min_capacity>
+using ethernet_write_pipe =
+  kernel_writeable_io_pipe<ethernet_pipe_id<1>, dataT, min_capacity>;
+} // namespace intelfpga */
+
+template <class name, class dataT, size_t min_capacity = 0>
+class kernel_readable_io_pipe {
+public:
+  // Non-blocking pipes
+  // Reading from pipe is lowered to SPIR-V instruction OpReadPipe via SPIR-V
+  // friendly LLVM IR.
+  static dataT read(bool &Success) {
+#ifdef __SYCL_DEVICE_ONLY__
+    RPipeTy<dataT> RPipe =
+      __spirv_CreatePipeFromPipeStorage_read<dataT>(&m_Storage);
+    dataT TempData;
+    Success = static_cast<bool>(
+        __spirv_ReadPipe(RPipe, &TempData, m_Size, m_Alignment));
+    return TempData;
+#else
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+  // Blocking pipes
+  // Reading from pipe is lowered to SPIR-V instruction OpReadPipe via SPIR-V
+  // friendly LLVM IR.
+  static dataT read() {
+#ifdef __SYCL_DEVICE_ONLY__
+    RPipeTy<dataT> RPipe =
+      __spirv_CreatePipeFromPipeStorage_read<dataT>(&m_Storage);
+    dataT TempData;
+    __spirv_ReadPipeBlockingINTEL(RPipe, &TempData, m_Size, m_Alignment);
+    return TempData;
+#else
+    assert(!"Pipes are not supported on a host device!");
+#endif // __SYCL_DEVICE_ONLY__
+  }
+
+private:
+  static constexpr int32_t m_Size = sizeof(dataT);
+  static constexpr int32_t m_Alignment = alignof(dataT);
+  static constexpr int32_t m_Capacity = min_capacity;
+  static constexpr int32_t ID = name::id;
+#ifdef __SYCL_DEVICE_ONLY__
+  static constexpr struct ConstantPipeStorage
+  m_Storage __attribute__((io_pipe_id(ID))) = {m_Size, m_Alignment, m_Capacity};
```

Comment:
Also for such a simple thing, just recycling a Clang ``annotate`` attribute could be enough, as a kind of Über attribute.

Reply:
That's a good suggestion. Not sure if annotation attribute can parse template parameters out of a box, but it worth a try.

Reply:
I made some experiments about this in August 2019 and forgot about it.
I have just pushed the current state at that time in https://github.com/keryell/teaching_C-plus-plus/blob/master/Clang/Annotate/annotate.cpp

Reply:
Since you cannot add values (template or not), the idea is to use for example some intermediate wrapping class where you store the templated or not value with some attributes on it.

Reply:
Sorry for a time without response - I was on a vacation without having the ability to build LLVM in a reasonable amount of time :)

So, I made some experiments here. The first one is kinda naive (https://godbolt.org/z/BHZh6g) - I just tried to add "pipe_id" annotation string as:
`static constexpr StorageTy Storage __attribute__((annotate("pipe_id"))) = { 42 };`
and I found out, that no annotation string was generated. It can be fixed by just adding "-std=c++2a" (https://godbolt.org/z/ghvpub), but, you know, we are still waiting for C++17... Also, I believe, that in CodeGenModule.cpp a code like:
```
if (D->hasAttr<AnnotateAttr>())
  AddGlobalAnnotations(D, GV);
```
is missing in a `GetOrCreateLLVMGlobal ` function definition and that possible can fix the issue as well.

I didn't really get the idea of the wrappers addition. Do I understand correctly, that you are suggesting to add a check (in backend) for to where the annotation string is attached, get inners of this structure and as this wrapper is used inside of the actual io_readable_pipe class - bound the constexpr ID value with the io pipe storage? Well, in theory it can work, but it looks even more overloaded than demangling approach.
I also tried to compile the example provided by you and saw, that there is not "bits" string generated for AP integer structure.

A little comment out of scope of IO pipe feature. Replacing all the attributes with annotation attributes is facing several issues: first one is coming from clang - we can't generate the appropriate diagnostics in case of incorrect usage of an attribute. Second one is Intel's SYCL specific (SPIR-V specific) - in SPIR-V there is no representation of the attribute (no appropriate decoration). We have added some extensions to overcome it, but I would really love to have these decorations in the core SPIR-V spec. Yet in general I'm down with reusing this Über attribute™ since with it we will no longer overstretch clang with these tons of vendor specific attributes.

Reply:
Interesting that it works in C++20 but not before...
But on the other hand, in triSYCL we did not use any attribute at all at the end because all the information we were looking was actually in the mangled name and we did not have the resource to develop a big front-end.
For example in your https://godbolt.org/z/BHZh6g example:
`_ZN4pipeI8ethernetILi1EEE4readEv` has directly the model of the pipe in it!
On UNIX
```
c++filt _ZN4pipeI8ethernetILi1EEE4readEv
pipe<ethernet<1> >::read()
```
Thus it depends on the ABI to be handled in the middle-end instead of depending on some front-end changes + middle-end.
But this is a completely different design compared to the current one used here, so this is purely informational about how to do cheap developments. :-)